File: agent-runner/internal/protocol/protocol.go
// Package protocol definisce i tipi di frame scambiati col gateway.
// Vedi docs/PROTOCOL.md per la specifica completa.
package protocol

const (
	Hello       = "hello"
	Welcome     = "welcome"
	Ping        = "ping"
	Pong        = "pong"
	TaskStart   = "task.start"
	TaskStarted = "task.started"
	ChatSend    = "chat.send"
	ChatDelta   = "chat.delta"
	ChatResult  = "chat.result"
	TaskDone    = "task.done"
	Error       = "error"

	RoleAgent = "agent"
)

// Inbound e' l'envelope generico ricevuto dal gateway: si legge prima Type, poi
// si accede ai campi pertinenti.
type Inbound struct {
	Type         string `json:"type"`
	SessionID    string `json:"session_id"`
	ProjectID    int    `json:"project_id"`
	TicketID     int    `json:"ticket_id"`
	BlobID       string `json:"blob_id"`
	Instructions string `json:"instructions"`
	Text         string `json:"text"`
	TS           int64  `json:"ts"`
	Code         string `json:"code"`
	Message      string `json:"message"`
}

// --- costruttori dei frame in uscita (verso il gateway) ---

func Hello_(agentID, token string, projects []map[string]any) map[string]any {
	return map[string]any{
		"type": Hello, "role": RoleAgent, "agent_id": agentID,
		"token": token, "projects": projects,
	}
}

func PingFrame(ts int64) map[string]any { return map[string]any{"type": Ping, "ts": ts} }
func PongFrame(ts int64) map[string]any { return map[string]any{"type": Pong, "ts": ts} }

func TaskStartedFrame(sessionID string, ticketID int, claudeSessionID, workdir string) map[string]any {
	return map[string]any{
		"type": TaskStarted, "session_id": sessionID, "ticket_id": ticketID,
		"claude_session_id": claudeSessionID, "workdir": workdir,
	}
}

func ChatResultFrame(sessionID, text string, isError bool, claudeSessionID string, costUSD float64, durationMS int64) map[string]any {
	return map[string]any{
		"type": ChatResult, "session_id": sessionID, "text": text, "is_error": isError,
		"claude_session_id": claudeSessionID, "cost_usd": costUSD, "duration_ms": durationMS,
	}
}

func ErrorFrame(sessionID, code, message string) map[string]any {
	return map[string]any{"type": Error, "session_id": sessionID, "code": code, "message": message}
}

File: gateway/soltea_gateway/protocol.py
"""Costanti e helper per i frame del protocollo (vedi docs/PROTOCOL.md)."""
from __future__ import annotations

from typing import Any

# --- tipi di frame ---
HELLO = "hello"
WELCOME = "welcome"
PING = "ping"
PONG = "pong"
RESOLVE_PROJECT = "resolve_project"
PROJECT_RESOLVED = "project_resolved"
LIST_AGENTS = "list_agents"
AGENTS = "agents"
TASK_START = "task.start"
TASK_STARTED = "task.started"
CHAT_SEND = "chat.send"
CHAT_DELTA = "chat.delta"
CHAT_RESULT = "chat.result"
TASK_DONE = "task.done"
ERROR = "error"

ROLE_AGENT = "agent"
ROLE_ORCHESTRATOR = "orchestrator"

# --- codici d'errore ---
ERR_UNAUTHORIZED = "unauthorized"
ERR_BAD_HELLO = "bad_hello"
ERR_NO_AGENT = "no_agent_for_project"
ERR_PROJECT_NOT_DECLARED = "project_not_declared"
ERR_UNKNOWN_SESSION = "unknown_session"
ERR_BLOB_NOT_FOUND = "blob_not_found"
ERR_CLAUDE_FAILED = "claude_failed"
ERR_INTERNAL = "internal"


def error(code: str, message: str, **extra: Any) -> dict[str, Any]:
    frame = {"type": ERROR, "code": code, "message": message}
    frame.update({k: v for k, v in extra.items() if v is not None})
    return frame
